### ISSUE TITLE
`hub clone public/repo` should not prompt for authentication

### ISSUE THREAD

**Issue by nelhage** [github/hub#1839]
I commonly drop `hub` onto a new laptop just for the `hub clone user/project` syntax. It'd be awesome if that usage didn't require authentication for public repositories. In fact, I'd ideally rather it didn't prompt for authentication at all and just translated the URL; I configure `git` to use `ssh` for all github repositories and so my ssh keys will silently kick in and allow the clone to succeed.

**Comment by mislav** [comment]
This has come up multiple times and I've dismissed it in the past, but by now it's clear that this is a shared pain. I agree that for public repos, whatever API requests are being made could be made using the public API without authentication. However, `hub clone user/project` would fail if `user/project` is a private repo. Would this be acceptable behavior?

**Comment by tedmiston** [comment]
My use case is the same as OP.

I came here to report the same issue after I couldn't figure out why cloning a private repo like `git clone git@github.com:myorg/myrepo.git` works fine but `hub clone myorg/myrepo` fails (prompting for authentication).

I attempted to leave the authentication prompts blank, give them bogus creds, etc to see if it would fallback to SSH key authentication but no dice.

Is it not possible to `hub clone ...` on private repos today?

**Comment by tedmiston** [comment]
I also saw the [`HUB_PROTOCOL`](https://github.com/github/hub/blob/2d4b6cb6b980b9e70a9b11b385dc1331d5faf69f/share/man/man1/hub.1.ronn#L138) env var mentioned in issue https://github.com/github/hub/issues/1788.

I gave that a shot but ran into the same issue:

```
$ export HUB_PROTOCOL=git
$ hub clone myrepo/myorg
github.com username: ^C
```

**Comment by mislav** [comment]
@tedmiston See https://github.com/github/hub/issues/1644#issuecomment-359002547

**Comment by tedmiston** [comment]
Thanks - I don't think it's clear to me why we would need to hit the GitHub API to clone.  Couldn't it simply be building the `git clone ...` command as a string from the input and running that directly?

I do this already in other scripts by generating a personal access token.

**Comment by tedmiston** [comment]
That makes sense.  I do want to communicate with the GitHub API on operations that require it.  I would strongly prefer not to grant access to that via my raw creds... I also have 2FA on my account and couldn't find documented whether that was supported.

I found in the docs that it's possible to bypass the username and password prompt by alternatively [providing an access token](https://github.com/github/hub/blame/3c8b674faa63584b23e75f0b0c05bb949f293013/share/man/man1/hub.1.ronn#L101-L102) via the `GITHUB_TOKEN` env var.

Is it possible to store this token somewhere else besides e.g., a global env var in my ~/.bash_profile?  Like git config, `~/.config/hub`, etc?  Alternatively, is there a way to receive a prompt that accepts the access token, instead of username & password, and takes core of storing it?

**Comment by mislav** [comment]
> I also have 2FA on my account and couldn't find documented whether that was supported.

2FA is supported. After supplying username and password, you will be prompted for 2FA code.



> I would strongly prefer not to grant access to that via my raw creds...

Sure, that makes sense. FWIW, hub doesn't send nor store your password anywhere else other than using it to log into `https://api.github.com` one time.

> Is it possible to store this token somewhere else besides e.g., a global env var

Yes. You can paste your Personal Access Token instead of your password after being prompted for username on the command line.

**Comment by x-yuri** [comment]
The way it works now would be tolerable (?) if it was documented. The [getting started][1] page apparently covers cloning public repositories, but it doesn't mention authentication one bit. Authentication is covered by the [man page][2] to a degree. But I just tried using `hub` on another laptop, it asked for username/password, then said, "Not Found." 2FA? On my old laptop it worked. It took me a while to figure out that I can store the Personal access token in `~/.config/hub`. An undocumented feature? A deprecated feature?

[1]: https://hub.github.com/
[2]: https://hub.github.com/hub.1.html

**Comment by iandunn** [comment]
I had the same problem as @x-yuri. Here's an example of what `~/.config/hub` should look like:

```
github.com:
- user: {your username}
  oauth_token: {personal token from https://github.com/settings/tokens}
```

**Comment by iandunn** [comment]
Storing a permanent token in a plain text file doesn't feel very secure, though. Is there a better way?

**Comment by dserodio** [comment]
> Storing a permanent token in a plain text file doesn't feel very secure, though. Is there a better way?

Yeah, I don't like it either. I don't know if `~/.config/hub` format is documented, but if this config could be read from `~/.gitconfig` instead, it would be possible to store the token in macOS Keychain (or similar for other OS) and dynamically read, as suggested [here](https://mumble.org.uk/blog/1/01/01/dotfiles-and-github/) since `git-config` can exec scripts to fetch this value.
